spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/ops/DataFlowOps.scala

```diff
@@ -0,0 +1,609 @@
+ *                               false.
+ * @tparam D Element numeric type in the tensor array.
+ */
+private[nn] class TensorArray[D: ClassTag](
+  private val initSize: Int,
+  private val shape: Array[Int] = null,
+  private var dynamicSize: Boolean = false,
+  private val clearAfterRead: Boolean = true,
+  private val identicalElementShapes: Boolean = false,
+  private val multipleWritesAggregate: Boolean = false
+)(implicit ev: TensorNumeric[D]) {
+
+  private var otherShape : Array[Int] = null
+
+  private var tensors = new Array[Tensor[D]](initSize)
+
+  def lockSize(): Unit = this.dynamicSize = false
+
+  def apply(index: Int): Tensor[D] = {
+    require(tensors(index) != null,
+      s"tensor on index $index has not been inited or has been cleared")
+    val t = tensors(index)
+    if (clearAfterRead) tensors(index) = null
+    t
+  }
+
+  def grad(): TensorArray[_] = {
+    this.lockSize()
+    new TensorArray[D](this.size, multipleWritesAggregate = true)
+  }
+
+  def size(): Int = tensors.length
+
+  def shapeOf(index: Int): Array[Int] = {
+    require(tensors(index) != null,
+      s"tensor on index $index has not been inited or has been cleared")
+    tensors(index).size()
+  }
+
+  def update(index: Int, tensor: Tensor[D]): Unit = {
+    if (!multipleWritesAggregate) {
+      require(tensors(index) == null, "There's already a tensor on the given index")
+    }
+
+    if (identicalElementShapes) {
+      if (otherShape == null) {
+        otherShape = tensor.size()
+      } else {
+        val curShape = tensor.size()
+        require(curShape.length == otherShape.length,
+          "insert tensor dimension does not match other tensor dimension")
+        var i = 0
+        while(i < curShape.length) {
+          require(curShape(i) == otherShape(i),
+            "insert tensor size does not match other tensor size")
+          i += 1
+        }
+      }
+    }
+    if (shape != null) {
+      val curShape = tensor.size()
+      require(curShape.length == shape.length,
+        "insert tensor dimension does not match required dimension")
+      var i = 0
+      while(i < curShape.length) {
+        require(curShape(i) == shape(i),
+          "insert tensor size does not match required size")
+        i += 1
+      }
+    }
+
+    if (dynamicSize && index >= tensors.size) {
+      val newTensors = new Array[Tensor[D]](index + 1)
+      var i = 0
+      while(i < tensors.length) {
+        newTensors(i) = tensors(i)
+        i += 1
+      }
+      tensors = newTensors
+    } else {
+      require(index < initSize, "cannot grow size when dynamicSize is false")
+    }
+
+    if (tensors(index) == null) {
+      tensors(index) = Tensor[D]().resizeAs(tensor).copy(tensor)
+    } else {
+      tensors(index).add(tensor)
+    }
+  }
+}
+
+private[nn] object TensorArray {
+  private val arrays = new ConcurrentHashMap[String, TensorArray[_]]()
```

Comment:
Do you think a WeakHashMap will solve the memory leak issue?

Reply:
I don't think it fit the requirement